src/Turing.jl

```diff
@@ -27,25 +28,72 @@ function setprogress!(progress::Bool)
 end
 
 # Log density function
-struct LogDensityFunction{V,M,S,C}
+struct LogDensityFunction{V,M,C}
     varinfo::V
     model::M
-    sampler::S
     context::C
 end
 
+function LogDensityFunction(
+    varinfo::DynamicPPL.AbstractVarInfo,
+    model::DynamicPPL.Model,
+    sampler::DynamicPPL.AbstractSampler,
+    context::DynamicPPL.AbstractContext
+)
+    return LogDensityFunction(varinfo, model, DynamicPPL.SamplingContext(sampler, context))
+end
+
+# Convenient for end-user.
+function LogDensityfunction(
+    model::DynamicPPL.Model,
+    varinfo::DynamicPPL.AbstractVarInfo=DynamicPPL.VarInfo(model),
+    context::DynamicPPL.AbstractContext=DynamicPPL.DefaultContext(),
+)
+    return LogDensityFunction(model, varinfo, context)
+end
+
+# HACK: heavy usage of `AbstractSampler` for, well, _everything_, is being phased out. In the mean time
+# we need to define these annoying methods to ensure that we stay compatible with everything.
+getsampler(f::LogDensityFunction) = getsampler(f.context)
+getsampler(ctx::DynamicPPL.SamplingContext) = ctx.sampler
+getsampler(ctx::DynamicPPL.AbstractContext) = getsampler(DynamicPPL.NodeTrait(ctx), ctx)
+getsampler(::DynamicPPL.IsParent, ctx::DynamicPPL.AbstractContext) = getsampler(DynamicPPL.childcontext(ctx))
+
+hassampler(f::LogDensityFunction) = hassampler(f.context)
+hassampler(ctx::DynamicPPL.SamplingContext) = true
+hassampler(ctx::DynamicPPL.AbstractContext) = hassampler(DynamicPPL.NodeTrait(ctx), ctx)
+hassampler(::DynamicPPL.IsLeaf, ctx::DynamicPPL.AbstractContext) = false
+hassampler(::DynamicPPL.IsParent, ctx::DynamicPPL.AbstractContext) = hassampler(DynamicPPL.childcontext(ctx))
+
+# TODO: This is annoying. Should all this be moved to DynamicPPL so we can share the impl of `unflatten`?
```

Comment:
My plan was to move the whole LogDensityFunction to DynamicPPL at some point, as it only depends on things defined therein and is not Turing-specific.

Reply:
I figured:) And as I added all these additional methods I was thinking "I'm probably going to just make a sibling-PR to DPPL with all of this". Given your comment too, I'll just do that :+1: